File: src/main/java/com/cuupa/dms/storage/StorageService.java
package com.cuupa.dms.storage;

import com.cuupa.dms.storage.document.Document;
import com.cuupa.dms.storage.document.db.DBDocument;
import com.cuupa.dms.storage.document.db.DocumentMapper;
import com.cuupa.dms.storage.document.db.MongoDBDocumentStorage;
import com.cuupa.dms.storage.tag.Tag;
import com.cuupa.dms.storage.tag.db.DBTag;
import com.cuupa.dms.storage.tag.db.MongoDBTagStorage;
import org.apache.commons.lang3.StringUtils;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

public class StorageService {

    private final MongoDBDocumentStorage documentStorage;

    private final MongoDBTagStorage tagStorage;

    public StorageService(MongoDBDocumentStorage documentStorage, MongoDBTagStorage tagStorage) {
        this.documentStorage = documentStorage;
        this.tagStorage = tagStorage;
    }

    public void deleteAll() {
        documentStorage.deleteAll();
        tagStorage.deleteAll();
    }

    public void save(Document document) {
        List<Tag> tags = document.getTags();
        tags.forEach(tag -> tag.setOwner(document.getOwner()));
        tags.forEach(tag -> tag.setName(tag.getName().trim()));
        List<DBTag> tagsInDB = getTagsFromDB(tags);

        saveNewTags(tags, tagsInDB);
        tagsInDB = getTagsFromDB(tags);
        System.out.println(tagsInDB);
        document.setTags(tags);
        DBDocument documentToSave = DocumentMapper.mapToEntity(document);

        documentStorage.insert(documentToSave);
    }

    private List<DBTag> getTagsFromDB(List<Tag> tags) {
        return tags.stream()
                   .map(tag -> tagStorage.findTagByNameAndOwner(tag.getName(), tag.getOwner()))
                   .filter(Objects::nonNull)
                   .collect(Collectors.toList());
    }


    private void saveNewTags(List<Tag> tags, List<DBTag> tagsInDB) {
        List<Tag>
                filteredTags =
                tags.stream()
                    .filter(tag -> !tagsInDB.stream()
                                            .map(tag1 -> tag.getName())
                                            .collect(Collectors.toList())
                                            .contains(tag.getName()))
                    .filter(tag -> !tagsInDB.stream()
                                            .map(tag1 -> tag.getOwner())
                                            .collect(Collectors.toList())
                                            .contains(tag.getOwner()))
                    .collect(Collectors.toList());


        List<DBTag>
                collect =
                filteredTags.stream().map(tag -> new DBTag(tag.getName(), tag.getOwner())).collect(Collectors.toList());
        for (DBTag tag : collect) {
            tagStorage.insert(tag);
        }
    }

    public List<Document> findDocumentsByOwner(String owner) {
        if (StringUtils.isBlank(owner)) {
            return new ArrayList<>();
        }
        List<DBDocument> documentsByOwner = documentStorage.findDocumentsByOwner(owner);

        List<Document> documents = new ArrayList<>(documentsByOwner.size());

        for (DBDocument document : documentsByOwner) {
            documents.add(DocumentMapper.mapToGuiObject(document));
        }

        return documents;
    }

    public List<Tag> findTagsByOwner(String owner) {
        return tagStorage.findTagsByOwner(owner).stream().map(tag -> new Tag(tag.getName(), tag.getOwner()))
                         .collect(Collectors.toList());
    }
}

File: src/main/java/com/cuupa/dms/storage/StorageService.kt
package com.cuupa.dms.storage

import com.cuupa.dms.service.CamundaService
import com.cuupa.dms.storage.document.Document
import com.cuupa.dms.storage.document.db.DBDocument
import com.cuupa.dms.storage.document.db.DocumentMapper
import com.cuupa.dms.storage.document.db.MongoDBDocumentStorage
import com.cuupa.dms.storage.tag.Tag
import com.cuupa.dms.storage.tag.db.DBTag
import com.cuupa.dms.storage.tag.db.MongoDBTagStorage
import java.util.*
import java.util.stream.Collectors

class StorageService(private val documentStorage: MongoDBDocumentStorage, private val tagStorage: MongoDBTagStorage,
                     private val camundaService: CamundaService) {

    fun deleteAll() {
        documentStorage.deleteAll()
        tagStorage.deleteAll()
    }

    fun save(document: Document): DBDocument {
        val tags = document.tags
        val tagsWithOwner = tags.map { tag: Tag -> tag.copy(owner = tag.owner) }
        val tagsInDB = getTagsFromDB(tagsWithOwner)
        saveNewTags(tags, tagsInDB)
        document.tags = tags

        val maxDocumentRevision = documentStorage.findDBDocumentByName(document.name).maxBy { it.revision }
        val documentToSave = DocumentMapper.mapToEntity(document)

        if (maxDocumentRevision != null) {
            documentToSave.revision = maxDocumentRevision.revision.inc()
        } else {
            documentToSave.revision = 1
        }
        return documentStorage.insert(documentToSave)
    }

    private fun getTagsFromDB(tags: List<Tag>): List<DBTag?> {
        return tags
                .map { tag: Tag -> tagStorage.findTagByNameAndOwner(tag.name, tag.owner) }
                .filter { obj: DBTag? -> Objects.nonNull(obj) }
    }

    private fun saveNewTags(tags: List<Tag>, tagsInDB: List<DBTag?>) {
        val filteredTags = tags.stream()
                .filter { tag: Tag ->
                    !tagsInDB.map { tag.name }
                            .contains(tag.name)
                }
                .filter { tag: Tag ->
                    !tagsInDB.map { tag.owner }
                            .contains(tag.owner)
                }
                .collect(Collectors.toList())
        val collect = filteredTags.map { tag: Tag -> DBTag(tag.name, tag.owner) }
        for (tag in collect) {
            tagStorage.insert(tag)
        }
    }

    fun findDocumentsByOwner(owner: String?): List<Document> {
        return if (owner.isNullOrBlank()) {
            listOf()
        } else findRevision(documentStorage.findDocumentsByOwnerOrderByName(owner).filterNotNull())
    }

    fun findDocumentsByOwnerAndProcessInstanceId(owner: String?, processInstanceIds: List<String>): List<Document> {
        return if (owner.isNullOrBlank()) {
            listOf()
        } else {
            findRevision(processInstanceIds.map { processInstanceId ->
                documentStorage
                        .findDBDocumentsByOwnerAndAndProcessInstanceId(owner, processInstanceId)
            }.filterNotNull())
        }
    }

    fun findTagsByOwner(owner: String): List<Tag> {
        return tagStorage.findTagsByOwner(owner)
                .stream()
                .map { tag: DBTag -> Tag(tag.name, tag.owner) }
                .collect(Collectors.toList())
    }

    fun complete(selectedDocument: Document?) {
        if (selectedDocument != null) {
            camundaService.complete(selectedDocument.processInstanceId)
            selectedDocument.processInstanceId = null
            save(selectedDocument)
        }
    }

    private fun findRevision(documents: List<DBDocument>): List<Document> {
        var lastDocument: DBDocument? = null
        val resultListFromDb: MutableList<DBDocument> = mutableListOf()
        for (document in documents) {
            if (lastDocument == null) {
                lastDocument = document
            }
            if (lastDocument.name == document.name) {
                if (document.revision > lastDocument.revision) {
                    lastDocument = document
                }
            } else {
                resultListFromDb.add(lastDocument)
                lastDocument = document
            }
        }
        if (!resultListFromDb.contains(lastDocument) && lastDocument != null) {
            resultListFromDb.add(lastDocument)
        }
        return resultListFromDb.map { document: DBDocument -> DocumentMapper.mapToGuiObject(document) }

    }
}